packages/core/reactive/context.ts

```diff
@@ -0,0 +1,49 @@
+interface InjectionConstraint<_> {}
+export type InjectionKey<T> = symbol & InjectionConstraint<T>
+
+export type Context = Map<symbol, unknown> & { parent?: Context }
+
+export const rootContext = createContext(true)
+let activeContext: Context = rootContext
+
+export function createContext(detached = false): Context {
+  const context = new Map() as Context
+  context.parent = detached ? void 0 : activeContext
+  return context
+}
+
+export function getCurrentContext(): Context {
+  return activeContext
+}
+
+export function runInContext<T>(context: Context, body: () => T): T {
+  const prevContext = activeContext
+  activeContext = context
+  try {
+    return body()
+  }
+  finally {
+    activeContext = prevContext
+  }
+}
+
+function lookup<T>(key: InjectionKey<T>): Context | undefined {
+  let context: Context | undefined = activeContext
+  while (context != null && !context.has(key)) {
```

Comment:
The `context` can never be `null`, maybe you want to say `undefined`. However, you could write this instead:
```ts
while (context && !context.has(key))
```